File: backend/app/schemas/bets.py
"""
Schemas: Bets
"""
from datetime import datetime
from enum import Enum
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.sessions import BankrollStrategy


class BetStatus(str, Enum):
    pending = "pending"
    resolved = "resolved"
    # Destino de una apuesta que sigue pending cuando se cierra la sesión,
    # para que no quede huérfana sin estado.
    cancelled = "cancelled"


# ---------- Requests ----------

class CreateBetRequest(BaseModel):
    category: str          # debe existir en game_variant.categories_json
    option_label: str      # debe existir dentro de esa categoría
    amount: float = Field(gt=0)
    followed_suggestion: bool = False
    # Gestión con la que se apostó. Solo esa progresión avanza de escalón al
    # resolverse el giro; null en una apuesta manual por fuera de ellas.
    strategy: Optional[BankrollStrategy] = None


# ---------- Responses ----------

class BetResponse(BaseModel):
    id: UUID
    session_id: UUID
    spin_id: Optional[UUID] = None
    category: str
    option_label: str
    amount: float
    followed_suggestion: bool
    strategy: Optional[BankrollStrategy] = None
    status: BetStatus
    won: Optional[bool] = None
    payout: Optional[float] = None
    net_change: Optional[float] = None
    created_at: datetime
    resolved_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class BetResolution(BaseModel):
    """Sub-objeto embebido en la respuesta de POST /spins cuando resuelve una apuesta pendiente."""
    bet_id: UUID
    category: str
    option_label: str
    won: bool
    amount: float
    payout: float
    net_change: float
    followed_suggestion: bool

File: frontend/lib/types/bets.ts
/**
 * Espejo manual de `backend/app/schemas/bets.py`.
 * Cambios en el schema Pydantic se replican aquí en el mismo commit
 * (skill `api-schema-sync`).
 */

import type { UUID } from "./auth";

export type BetStatus = "pending" | "resolved" | "cancelled";

export interface CreateBetRequest {
  category: string;
  option_label: string;
  amount: number;
  /**
   * Nota del usuario sobre su propia decisión: si eligió esta opción después de
   * ver una señal. No implica que la señal anticipara el resultado.
   */
  followed_suggestion: boolean;
}

export interface BetResponse {
  id: UUID;
  session_id: UUID;
  spin_id: UUID | null;
  category: string;
  option_label: string;
  amount: number;
  followed_suggestion: boolean;
  status: BetStatus;
  won: boolean | null;
  /** Total devuelto por la mesa: lo apostado más la ganancia. Cero si se perdió. */
  payout: number | null;
  net_change: number | null;
  created_at: string; // ISO 8601
  resolved_at: string | null;
}

export interface BetResolution {
  bet_id: UUID;
  category: string;
  option_label: string;
  won: boolean;
  amount: number;
  payout: number;
  net_change: number;
  followed_suggestion: boolean;
}
